### PR TITLE
Add support for graceful coordinator shutdown

### PR DESCRIPTION
**Pull request by wendigo**
Release notes: add coordinator graceful shutdown support

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        log.info("Shutting down");
+        List<BasicQueryInfo> activeQueries = getActiveQueries();
+
+        // At this point no new queries should be queued by the coordinator.
+        // Wait for all remaining queries to finish.
+        while (!activeQueries.isEmpty()) {
+            CountDownLatch countDownLatch = new CountDownLatch(activeQueries.size());
+
+            for (BasicQueryInfo queryInfo : activeQueries) {
+                queryManager.addStateChangeListener(queryInfo.getQueryId(), newState -> {
```
Comment (nineinchnick) [inline core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java:99]: Is a race condition possible here, when an active query will be finished before attaching this change listener?

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java:99]: no, addStateChangeListener will always fire at least once with the last observed state afair

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java:99]:
```
// fire state change listener with the current state
        // always fire listener callbacks from a different thread
        safeExecute(() -> fireStateChangedListener(currentState, stateChangeListener));
```

File: core/trino-main/src/main/java/io/trino/dispatcher/QueuedStatementResource.java [hunk 2]
```diff
@@ -170,6 +175,10 @@ public Response postStatement(
             throw badRequest(BAD_REQUEST, "SQL statement is empty");
         }
 
+        if (status.isShutdownStarted()) {
+            return badRequest(GONE, "Coordinator is shutting down").getResponse();
```
[the remark this answers is not in the document]
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/dispatcher/QueuedStatementResource.java]: Done

File: core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+                    if (newState.isDone()) {
+                        countDownLatch.countDown();
+                    }
+                });
+            }
+
+            log.info("Waiting for all queries to finish, %d still active", countDownLatch.getCount());
+
+            try {
+                countDownLatch.await();
```
[the remark this answers is not in the document]
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java]: Not anymore

File: core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+    public synchronized void requestShutdown()
+    {
+        log.info("Shutdown requested");
+        if (shutdownRequested) {
+            return;
+        }
+        shutdownRequested = true;
+
+        status.shutdownStarted();
+        // wait for a grace period to start the shutdown sequence
```
[the remark this answers is not in the document]
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java:86]: Changed

**Comment by wendigo** [comment]
@kokosing @nineinchnick PTAL

File: client/trino-client/src/main/java/io/trino/client/ServerInfo.java [hunk 5]
```diff
@@ -41,12 +42,14 @@ public ServerInfo(
             @JsonProperty("environment") String environment,
             @JsonProperty("coordinator") boolean coordinator,
             @JsonProperty("starting") boolean starting,
+            @JsonProperty("shutting") boolean shutting,
```
Comment (nineinchnick) [inline client/trino-client/src/main/java/io/trino/client/ServerInfo.java:45]: Nit: how about `stopping`?

File: core/trino-main/src/main/java/io/trino/server/ServerConfig.java [hunk 6]
```diff
@@ -104,4 +121,10 @@ public ServerConfig setQueryInfoUrlTemplate(String queryInfoUrlTemplate)
         this.queryInfoUrlTemplate = Optional.ofNullable(queryInfoUrlTemplate);
         return this;
     }
+
+    public enum CoordinatorShutdownStrategy
+    {
+        PREEMPT,
```
Comment (nineinchnick) [inline core/trino-main/src/main/java/io/trino/server/ServerConfig.java:127]: These values are very vague, how about `CANCEL_QUERIES` and `WAIT_FOR_RUNNING_QUERIES`?

File: core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    {
+        List<QueryId> activeQueries = getActiveQueryIds();
+
+        try {
+            switch (strategy) {
+                case WAIT_FOR_COMPLETION -> checkState(waitForQueriesCompletion(activeQueries, Duration.ZERO), "Not all queries completed");
+                case CANCEL_RUNNING -> cancelRunningQueries(activeQueries);
+            }
+
+            Thread.sleep(10_000); // wait 10s for the clients to receive last result
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java:104]: should we taking this from some config or constant instead of hardcoding value?

File: core/trino-main/src/main/java/io/trino/server/GracefulShutdownWorkerHandler.java [hunk 8]
```diff
@@ -0,0 +1,153 @@
...
+        sleepUninterruptibly(gracePeriod.toMillis(), MILLISECONDS);
+
+        Future<?> shutdownFuture = lifeCycleStopper.submit(() -> {
+            lifeCycleManager.stop();
+            return null;
+        });
+
+        // terminate the jvm if life cycle cannot be stopped in a timely manner
+        try {
+            shutdownFuture.get(LIFECYCLE_STOP_TIMEOUT.toMillis(), MILLISECONDS);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/server/GracefulShutdownWorkerHandler.java:130]: I thought the outside task is doing the same. It calls graceful shutdown and waits till completion for some time, if it is still up and running it kills-nine it

**Review by kokosing — COMMENTED** [review]
don't we need to update any scripts that calls graceful shutdown now? What is the point of changing the output if nobody consumes it?

skimmed

File: core/trino-main/src/main/java/io/trino/server/GracefulShutdownHandler.java [hunk 9]
```diff
@@ -13,145 +13,9 @@
...
-import static com.google.common.util.concurrent.Uninterruptibles.sleepUninterruptibly;
-import static io.airlift.concurrent.Threads.threadsNamed;
-import static java.lang.Thread.currentThread;
-import static java.util.Objects.requireNonNull;
-import static java.util.concurrent.Executors.newSingleThreadExecutor;
-import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
-import static java.util.concurrent.TimeUnit.MILLISECONDS;
-import static java.util.concurrent.TimeUnit.SECONDS;
-
-public class GracefulShutdownHandler
```
Comment (electrum) [inline core/trino-main/src/main/java/io/trino/server/GracefulShutdownHandler.java:42]: Please do this refactoring in a separate commit

**Review by electrum — DISMISSED** [review]

File: core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+            return;
+        }
+        shutdownRequested = true;
+
+        // mark coordinator as shutting down - new query submissions will be rejected
+        status.shutdownStarted();
+        log.info("Coordinator is now rejecting new query submissions");
+
+        // wait for a grace period to start the shutdown sequence
+        shutdownHandler.schedule(this::shutdown, gracePeriod.toMillis(), MILLISECONDS);
```
[the remark this answers is not in the document]
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/server/GracefulShutdownCoordinatorHandler.java:91]: I've followed the same logic for worker shut down sequence. Maybe it's not needed for coordinator since we've marked as shutting down which will reject new query submissions

**Comment by mosabua** [comment]
General high level question from a docs perspective .. Here is what I think should happen

* Coordinator shutdown is requested
* No more new queries are accepted
* Existing queries drain and finish via graceful shutdown requests to all workers
* All workers shutdown
* Coordinator shuts down

Btw .. this will be a great feature to have and potentially integrate into the Trino Gateway

**Comment by mosabua** [comment]
> @mosabua coordinator won't be responsible for a worker shutdown, this should be initiated by the cluster manager, like K8S

Why? That task is literally in the name of the "coordinator" .. coordinate the cluster.

Whats the point of a coordinator shutdown if it potentially leaves workers dangling?

**Comment by wendigo** [comment]
@mosabua depends on the configurable strategy. Either wait for queries or cancel them if they didn't finish in a specified time

**Comment by kokosing** [comment]
> That task is literally in the name of the "coordinator" .. coordinate the cluster.

@mosabua I don't agree. "coordinator" is more about query processing coordination. One could imagine that in future cluster may have multiple coordinators if that becomes a bottleneck. I believe each node should control its life cycle independently. Like @wendigo said.

**Comment by mosabua** [comment]
> > That task is literally in the name of the "coordinator" .. coordinate the cluster.
> 
> @mosabua I don't agree. "coordinator" is more about query processing coordination. One could imagine that in future cluster may have multiple coordinators if that becomes a bottleneck. I believe each node should control its life cycle independently. Like @wendigo said.

I can see both sides of the argument. However it looks like with recent improvements like dynamic catalogs the vision is that the coordinator does more than just coordinate query processing. How do you think of the role of the coordinator @electrum and @dain ?

**Comment by wendigo** [comment]
@mosabua if it's not coordinated with the cluster manager like k8s, worker nodes going down will be restarted back, so the proper flow of control is the other way around than you are suggesting

**Comment by wendigo** [comment]
@mosabua this task is about making coordinator go down when asked by k8s, right now it will wait for grace period and be killed, anything other than that is an improvement

**Comment by mosabua** [comment]
> @mosabua this task is about making coordinator go down when asked by k8s, right now it will wait for grace period and be killed, anything other than that is an improvement

Sounds good .. I do believe we should improves this over time. We definitely have to document how it currently works. I assume you will do that in this PR ...

File: core/trino-main/src/main/java/io/trino/server/ServerConfig.java [hunk 11]
```diff
@@ -104,4 +121,10 @@ public ServerConfig setQueryInfoUrlTemplate(String queryInfoUrlTemplate)
         this.queryInfoUrlTemplate = Optional.ofNullable(queryInfoUrlTemplate);
         return this;
     }
+
+    public enum CoordinatorShutdownStrategy
+    {
+        CANCEL_RUNNING, // Cancel running queries after grace period has passed
+        ABORT_AFTER_WAITING // Wait for running queries to complete and abort shutdown after grace period has passed
```
Comment (nineinchnick) [inline core/trino-main/src/main/java/io/trino/server/ServerConfig.java:128]: That doesn't sound very useful, a busy cluster would never get shut down. Why there's no longer an option to wait, but then cancel all remaining queries?

File: core/trino-main/src/main/java/io/trino/server/ServerConfig.java [hunk 12]
```diff
@@ -104,4 +121,10 @@ public ServerConfig setQueryInfoUrlTemplate(String queryInfoUrlTemplate)
         this.queryInfoUrlTemplate = Optional.ofNullable(queryInfoUrlTemplate);
         return this;
     }
+
+    public enum CoordinatorShutdownStrategy
+    {
+        CANCEL_RUNNING, // Cancel running queries after grace period has passed
```
Comment (nineinchnick) [inline core/trino-main/src/main/java/io/trino/server/ServerConfig.java:127]: The name is misleading, I initially thought it cancels all running queries immediately. Should there be a third option to do this? It would inform clients the server is either being restarted so they can retry, or shut down so they don't have to keep trying.

File: core/trino-main/src/main/java/io/trino/server/ServerConfig.java (hunk 11 above)
Comment (nineinchnick) [inline core/trino-main/src/main/java/io/trino/server/ServerConfig.java:128]: This name is misleading too, it reads as if the shutdown is _always_ aborted. How about `WAIT_AND_CANCEL_QUERIES` and `WAIT_AND_ABORT_SHUTDOWN`?

**Comment by oneonestar** [comment]
I would like to propose the following design:

### Shutting down (accept new query)
Add a `Shutting down (accept new query)` period after receive shutdown request.
During this period, "/info/state" and "/status" will indicate the coordinator is shutting down.
This allows external system like trino-gateway to pickup the status and stop routing query to coordinator.

### Shutting down (reject new query)
After grace-period, coordinator will transit to `Shutting down (reject new query)`.
Coordinator can wait for "some condition" depends on the configuration, such as:
* `WAIT_FOR_ALL_QUERY_COMPLETE`: Wait until all queries are completed without timeout
* `WAIT_FOR_GRACE_PERIOD`: Wait until all queries are completed with timeout = grace_period
* `CANCEL_RUNNING`: Kill all running queries immediately
* `ABORT_AFTER_WAITING`: Wait for running queries to complete and abort shutdown after grace period has passed


### FINAL_WAITING
Finally, coordinator will wait for another grace-period in `FINAL_WAITING` phase.
This allows coordinator to clean up, such as allowing event listener to send the last message or allowing clients to receive final response.


```mermaid
stateDiagram-v2
    ACTIVE --> SHUTTING_DOWN_PHASE_1: receive shutdown request
    SHUTTING_DOWN_PHASE_1 --> SHUTTING_DOWN_PHASE_2: wait for grace-period
    SHUTTING_DOWN_PHASE_2 --> FINAL_WAITING: wait for "some condition"
    SHUTTING_DOWN_PHASE_2 --> ACTIVE: ABORT_AFTER_WAITING
    FINAL_WAITING --> TERMINATE: wait for grace-period

    SHUTTING_DOWN_PHASE_1: Shutting down (accept new query)
    SHUTTING_DOWN_PHASE_2: Shutting down (reject new query)
```

**Comment by wendigo** [comment]
@oneonestar first state can be implemented in the gateway itself and it's redundant from the k8s deployment pov. I'd like to avoid having a state machine for the shutdown process

**Comment by oneonestar** [comment]
@wendigo 
Most of the managed k8s platform including GKE and EKS have a node auto-upgrade feature. Some arbitrary nodes would be selected and drained for an upgrade. It’s hard to notify the gateway without first state.

**Comment by mosabua** [comment]
👋 @wendigo  .. I assume you will rebase and continue to run with this PR.

**Comment by wendigo** [comment]
Correct @mosabua

**Comment by mosabua** [comment]
Just a reminder @wendigo  ;-)
